server/src/main/java/org/elasticsearch/index/engine/ThreadPoolMergeExecutor.java

```diff
@@ -0,0 +1,156 @@
+                return 1;
+            } else if (mergeTask2 == null) {
+                // "merge task 1" can run because "merge scheduler 2" cannot run any merges
+                return -1;
+            } else {
+                // run smaller merge task first
+                return mergeTask1.compareTo(mergeTask2);
+            }
+        }
+    });
+    private final ExecutorService executorService;
+    private final int maxConcurrentMerges;
+    private int currentlyExecutingMergesCount;
+    private int currentlyActiveIOThrottledMergesCount;
+
+    public ThreadPoolMergeExecutor(ThreadPool threadPool) {
+        this.executorService = threadPool.executor(ThreadPool.Names.MERGE);
+        this.maxConcurrentMerges = threadPool.info(ThreadPool.Names.MERGE).getMax();
+    }
+
+    public double getTargetMBPerSec() {
+        return targetMBPerSec;
+    }
+
+    public synchronized void registerMergeScheduler(ThreadPoolMergeScheduler threadPoolMergeScheduler) {
+        if (registeredMergeSchedulers.add(threadPoolMergeScheduler) == false) {
+            throw new IllegalStateException("cannot register the same scheduler multiple times");
+        }
+    }
+
+    public synchronized void unregisterMergeScheduler(ThreadPoolMergeScheduler threadPoolMergeScheduler) {
+        if (registeredMergeSchedulers.remove(threadPoolMergeScheduler) == false) {
+            throw new IllegalStateException("cannot unregister if the scheduler has not been registered");
+        }
+    }
+
+    public synchronized void updateMergeScheduler(
+        ThreadPoolMergeScheduler threadPoolMergeScheduler,
+        Consumer<ThreadPoolMergeScheduler> updater
+    ) {
+        boolean removed = registeredMergeSchedulers.remove(threadPoolMergeScheduler);
+        if (false == removed) {
+            throw new IllegalStateException("Cannot update a merge scheduler that is not registered");
+        }
+        currentlyExecutingMergesCount -= threadPoolMergeScheduler.getCurrentlyRunningMergeTasks().size();
+        currentlyActiveIOThrottledMergesCount -= getIOThrottledMergeTasksCount(threadPoolMergeScheduler);
+        updater.accept(threadPoolMergeScheduler);
+        boolean added = registeredMergeSchedulers.add(threadPoolMergeScheduler);
+        if (false == added) {
+            throw new IllegalStateException("Found duplicate registered merge scheduler");
+        }
+        currentlyExecutingMergesCount += threadPoolMergeScheduler.getCurrentlyRunningMergeTasks().size();
+        currentlyActiveIOThrottledMergesCount += getIOThrottledMergeTasksCount(threadPoolMergeScheduler);
+        double newTargetMBPerSec = maybeUpdateTargetMBPerSec();
+        if (newTargetMBPerSec != targetMBPerSec) {
+            targetMBPerSec = newTargetMBPerSec;
+            threadPoolMergeScheduler.setIORateLimitForAllMergeTasks(newTargetMBPerSec);
+        }
+        maybeExecuteNextMerges();
+    }
```

Comment:
An to a threadpool has the potential to reorder it in the sorted set.

Reply:
It also has the potential to change the stats related to IO throttling, because that is done "globally" i.e. across all shards on the node.